Otel trace groups

KarstenSchnitter opened the pull request:
### Description
Adjustment of the mapping templates for the standard OTel format.

### Check List
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

ps48 commented on data-prepper-plugins/opensearch/src/main/resources/index-template/otel-v1-apm-span-index-standard-template.json:
```diff
@@ -3,39 +3,136 @@
       "dynamic_templates": [
         {
-          "resource_attributes_map": {
+          "integer_resource_attributes": {
```
Should this be named as `long_resource_attributes` and similarly below as well

ps48 commented on data-prepper-plugins/opensearch/src/main/resources/index-template/logs-otel-v1-index-standard-template.json:
```diff
@@ -6,58 +6,129 @@
-        },
         "instrumentationScope": {
           "properties": {
+            "droppedEventsCount": {
+              "type": "integer"
```
Should this be `droppedAttributesCount` based on the OTel proto: https://github.com/open-telemetry/opentelemetry-proto/blob/d7770822d70c7bd47a6891fc9faacc66fc4af3d3/opentelemetry/proto/common/v1/common.proto#L80?

KarstenSchnitter replied:
Integer is at 32bits but signed. The proto definition specifies an unsigned 32bit integer. Do you think this difference is significant? I find it hard to imagine there would be a scenario where that number of attributes is ever attached and dropped to a log/metric/span.

KarstenSchnitter replied:
You are just referring to the naming, right? I am going to correct that.

vamsimanohar commented on data-prepper-plugins/opensearch/src/main/resources/index-template/logs-otel-v1-index-standard-template.json:
```diff
@@ -6,58 +6,129 @@
       },
+      "dynamic_templates": [
+        {
+          "integer_resource_attributes": {
```
shouldn't this be `long_resource_attributes`
I think the question is already raised in other comments. can we address that in every signal type.

KarstenSchnitter replied:
WIll do.

vamsimanohar commented on data-prepper-plugins/opensearch/src/main/resources/index-template/logs-otel-v1-index-standard-template.json:
```diff
@@ -6,58 +6,129 @@
-             "droppedEventsCount": {
-               "type": "integer"
-             }
+            "droppedEventsCount": {
```
shouldn't this be `droppedEventsCount`?

KarstenSchnitter replied:
I am confused. What do want to change? This is the name of the field. Only indentation seems to be off.

KarstenSchnitter replied:
I am going to change this to `droppedAttributesCount`.

vamsimanohar commented on data-prepper-plugins/opensearch/src/main/resources/index-template/logs-otel-v1-index-standard-template.json:
```diff
@@ -78,50 +149,56 @@
-          "ignore_above": 256,
-          "type": "keyword"
+          "type": "keyword",
+          "ignore_above": 16
```
Why are we limiting to 16 chars here and 32 in traceId?

KarstenSchnitter replied:
First, this is not a limitation of storage, just of the indexing.  [W3C Trace Context](https://www.w3.org/TR/trace-context/) specifies the trace id to be an array of 16 bytes and the span id to be an array of 8 bytes. With hex-encoding this produces 32 and 16 characters respectively. Using [B3 context](https://github.com/openzipkin/b3-propagation) would provide the same length. So it seems to be a reasonable choice.

vamsimanohar commented on data-prepper-plugins/opensearch/src/main/resources/index-template/logs-otel-v1-index-standard-template.json:
```diff
@@ -78,50 +149,56 @@
+          "type": "alias",
+          "path": "resource.attributes.service.name"
         },
         "event": {
```
Why do we need event here?
https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/logs/v1/logs.proto
We don't have anything in logs proto model.

KarstenSchnitter replied:
That is a question I am wondering myself. This comes out of the [Simple Schema for Observability](https://github.com/opensearch-project/opensearch-catalog/blob/main/docs/schema/observability/logs/README.md#event-fields). I kept it, since it was in the original file proposed by @kkondaka. I know this is not in the OpenTelemetry standard. The same is true for the `data_stream`. How do we want to go about this?

kkondaka replied:
I will remove this.

KarstenSchnitter replied:
Ok, I take it out in my PR to your branch.

vamsimanohar commented on data-prepper-plugins/opensearch/src/main/resources/index-template/logs-otel-v1-index-standard-template.json:
```diff
@@ -6,58 +6,129 @@
       "_source": {
         "enabled": true
```
can you also add eventName which is missing: https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/logs/v1/logs.proto#L226

We have done some analysis on logs mapping here: https://github.com/opensearch-project/opensearch-catalog/issues/217

KarstenSchnitter replied:
Will add the field. I am going to take a look at the analysis.

kkondaka replied:
@KarstenSchnitter It looks like we need to bump up the version number to 3 (at least for spans). Otherwise new template mappings are not being sent to the OpenSearch server.

KarstenSchnitter replied:
Where does this limitation come from?

vamsimanohar replied:
Have we added this?

KarstenSchnitter replied:
Also, this could look much nicer, if I could use component templates. That way, I could use the same dynamic mappings on all three templates. Is there a way to do that with Data Prepper?

vamsimanohar commented on data-prepper-plugins/opensearch/src/main/resources/index-template/logs-otel-v1-index-standard-template.json:
```diff
@@ -78,50 +149,56 @@
         },
         "flags": {
-          "type": "integer"
+          "type": "long"
```
https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/logs/v1/logs.proto#L178
Can we also add dropped_attributes_count in log

KarstenSchnitter replied:
This is defined as `fixed32`. Since this is interpreted as a bit field, I decided to use `long` to not change representation, when the highest bit is set. Do you prefer using `integer`, since only the lowest bits are currently in use?

KarstenSchnitter replied:
I am going to check, that for very attribute group there is a dropped attributes count. Thanks for pointing this out.

KarstenSchnitter replied:
I renamed all the `droppedEventsCount` to `droppedAttributesCount`.

vamsimanohar commented on data-prepper-plugins/opensearch/src/main/resources/index-template/logs-otel-v1-index-standard-template.json:
```diff
@@ -78,50 +149,56 @@
+                  }
                 },
                 "stacktrace": {
                   "type": "text"
```
```
        "@timestamp": {
          "type": "date_nanos"
        },
        "time": {
          "type": "date_nanos"
        },
        "observedTimestamp": {
          "type": "date_nanos"
        },
        "observedTime": {
          "type": "alias",
          "path": "observedTimestamp"
        },
        ```


        We have four fields related to timestamp. 
        We have two in otel proto files, can we consolidate? If they are required why?

KarstenSchnitter replied:
This was in the original file. The OTel specification only guarantees, that at least one of `time` and `observedTime` is set. It is not telling which one. I think, `@timestamp` is supposed to always contain a timestamp for each message. Furthermore, it provides a common timestamp field for all three signals. The alias seems to be there for compatibility.

kkondaka replied:
I think `@timestamp` is opensearch specific. Correct me if I am wrong. And `observedTime` is incorrect. I think `observedTimestamp` is correct. I do not think my new code generates "@timestamp" and "observedTime". The mappings have "observedTime" as alias for "observedTimestamp", which may be removed but I think it's ok to keep it.

KarstenSchnitter replied:
`Time` vs `Timestamp` is essentially OpenTelemetry [proto-files](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/logs/v1/logs.proto) vs [Logs Data Model](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-timestamp). I am fine with either naming as long as it is consistent.

vamsimanohar commented on data-prepper-plugins/opensearch/src/main/resources/index-template/otel-v1-apm-span-index-standard-template.json:
```diff
@@ -3,39 +3,136 @@
-        "enabled": true
-      },
       "properties": {
+        "droppedEventsCount": {
```
`dropperAttributesCount`. Also can you validate the existing mapping with 
https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/trace/v1/trace.proto

I see a lot of missing pieces.

vamsimanohar replied:
we did some analysis here: https://github.com/opensearch-project/opensearch-catalog/issues/218

KarstenSchnitter replied:
I am going to do that.

KarstenSchnitter commented:
I addressed the comments by @ps48 and @vamsimanohar. There are some open questions:

- What do we do about the events in the logs model? This is not part of the OTel Logs Data Model but in sso4. I kept the mappings, since they do not hurt when there is no data.
- Timestamp handling: I kept a common `@timestamp` field in logs, metrics and traces. The idea is to have a clear timestamp to align on and ease analysis across the signals. For logs `@timestamp` should contain the optional `time` and `observedTime` if it is not available. For metrics `@timestamp` should contain the `time`. We can make it an alias, but for index patterns it would be better to be at least a [copy_to](https://opensearch.org/docs/latest/field-types/mapping-parameters/copy-to/). Since it needs custom logic in the logging case, it is not an option there.

kkondaka commented on data-prepper-plugins/opensearch/src/main/resources/index-template/logs-otel-v1-index-standard-template.json:
```diff
@@ -6,58 +6,131 @@
+          }
+        },
+        {
+          "default_resource_attributes": {
```
@KarstenSchnitter Is this correct? Default attributes can be objects also, right? It looks like, here you are assuming them to be strings. I think it is possible for attributes to be nested like 
```
"resource": {
    "attributes" : {
          "data" : {
                 "key1": "value1"
                 "key2": 2
           },
           "key3": "value3"
      }
}
```

KarstenSchnitter replied:
Due to the `path_match` this is applied to all nested fields. With the three dynamic mappings your example would have the values of key1 indexed as keyword, of key2 as log, and of key3 as keyword.

kkondaka replied:
@KarstenSchnitter I am seeing this error with the mappings you provided - 
```
WARN  org.opensearch.dataprepper.plugins.sink.opensearch.BulkRetryStrategy - index = otel-v1-apm-span-000001, operation = Index, status = 400, error = object mapping for [resource.attributes.host.image.id] tried to parse field [host.image.id] as object, but found a concrete value
```

Any idea why this error is occurring?

The mappings from the server 
```
dynamic_templates" : [
        {
          "long_resource_attributes" : {
            "path_match" : "resource.attributes.*",
            "mapping" : {
              "type" : "long"
            },
            "match_mapping_type" : "long"
          }
        },
        {
          "double_resource_attributes" : {
            "path_match" : "resource.attributes.*",
            "mapping" : {
              "type" : "double"
            },
            "match_mapping_type" : "double"
          }
        },
        {
          "default_resource_attributes" : {
            "path_match" : "resource.attributes.*",
            "mapping" : {
              "type" : "object"
            }
          }
        },
        

```

kkondaka replied:
I have deleted the "default_resource_attributes", "default_scope_attributes" and "default_attributes" and then I was able to store the data and see the data using opensearch dashboards.

KarstenSchnitter replied:
This looks like a typical type mismatch in the error message. Note, that the `default_resource_attributes` is not what I provided. It should have type `keyword`. Do you have a procedure to reproduce this issue?

kkondaka replied:
@KarstenSchnitter sorry, that was copy-paste mistake. Even with the correct mappings with "keyword", I am still seeing the error. For now, I am going to take out the default attributes section in all three and updated the PR. These changes are working fine for traces,logs and metrics. Please submit another PR separately (on top of this) for any changes you want to make after I merge these change.

KarstenSchnitter replied:
Please do not take it out. Mapping to keywords is the essential default we need. Can you try with an additional `"match_mapping_type": "string"` instead? I have not seen this problem during my testing. Can you provide the events you were testing with?

kkondaka commented on data-prepper-plugins/opensearch/src/main/resources/index-template/logs-otel-v1-index-standard-template.json:
```diff
@@ -6,58 +6,131 @@
+          }
+        },
+        {
+          "default_attributes": {
```
Same comment here. General attributes may also have nested attributes, right?

KarstenSchnitter reviewed (commented):
I removed the event field from the logs mapping and answered some of the review comments. Timestamp names remain an open discussion.

kkondaka commented:
@KarstenSchnitter ignore my comment about version being 3. Will keep it as 1. If there is an existing template with version 1 already present, it needs to be removed. Otherwise, we have to bump this version 2.
